crates/core/tedge_write/tests/main.rs

```diff
@@ -0,0 +1,84 @@
+use std::fs;
+use std::fs::Permissions;
+use std::os::unix::fs::PermissionsExt;
+use std::path::PathBuf;
+
+use assert_cmd::Command;
+use tempfile::TempDir;
+
+#[test]
+fn creates_dest_file_if_doesnt_exist() {
+    // Arrange
+    let (temp_dir, source_path) = setup_source_file();
+    let destination_path = temp_dir.path().join("destination.txt");
+
+    let mut command = Command::cargo_bin(env!("CARGO_PKG_NAME")).unwrap();
+    command.pipe_stdin(&source_path).unwrap();
+    command.arg(&destination_path);
+
+    // Act
+    command.assert().success();
+
+    // Assert
+    assert_eq!(
+        fs::read_to_string(source_path).unwrap(),
+        fs::read_to_string(destination_path).unwrap(),
+    );
+}
+
+#[test]
+fn sets_file_permissions_if_file_doesnt_exist() {
+    // Arrange
+    let (temp_dir, source_path) = setup_source_file();
+    fs::set_permissions(&source_path, Permissions::from_mode(0o644)).unwrap();
+
+    let destination_path = temp_dir.path().join("destination.txt");
+
+    let mut command = Command::cargo_bin(env!("CARGO_PKG_NAME")).unwrap();
+    command.pipe_stdin(&source_path).unwrap();
+    command.arg(&destination_path);
+
+    command.args(["--mode", "600"]);
+
+    // Act
+    command.assert().success();
+
+    // Assert
+    let dest_mode = destination_path.metadata().unwrap().permissions().mode();
+    // .mode() returns st_mode, we only need to compare a subset
+    assert_eq!(dest_mode & 0o600, 0o600);
```

Comment:
I'm not convinced the bitwise `&` is achieving what you're looking for here. As far as I understand, you're receiving more bits than just the three you care about. I believe that you're trying to test that the 3 least-significant octal digits are `600`, but a quick test in irust shows that's not the case:

```
In: format!("{:o}", 0o704 & 0o600)
Out: "600"
```

I think what you want is:

```suggestion
    assert_eq!(dest_mode & 0o777, 0o600);
```

Reply:
You're completely right, that's a very dumb mistake.